File: android/app/src/main/java/gg/yappy/app/ui/chat/ChartView.kt
package gg.yappy.app.ui.chat

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.background
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import gg.yappy.app.data.EmbedChart
import gg.yappy.app.ui.theme.neuColors
import kotlin.math.abs

/**
 * Inline charts for embeds — the whole family, drawn by hand.
 *
 * No chart library: a message-sized chart needs exactly one series, no pan,
 * no zoom, no tooltips, and a look that matches the app rather than a
 * spreadsheet. Axis chrome is kept to a whisper (min/max, first/last label)
 * because in a chat bubble the *shape* is the message; anyone who wants the
 * numbers has them in the text beside it.
 */

private val SERIES = listOf(
    Color(0xFF8B7CFF), // violet
    Color(0xFF00CEC9), // teal
    Color(0xFFFF9F43), // orange
    Color(0xFFFF6B81), // pink
    Color(0xFFFCCE09), // yellow
    Color(0xFF6BCB77), // green
    Color(0xFF4FC3F7), // sky
    Color(0xFFB39DDB), // lavender
)

@Composable
fun ChartView(chart: EmbedChart, modifier: Modifier = Modifier) {
    val colors = neuColors
    val points = chart.points.filter { it.value.isFinite() }
    if (points.size < 2) return

    when (chart.kind) {
        "pie", "donut" -> PieChart(chart, donut = chart.kind == "donut", modifier)
        else -> {
            Column(modifier.fillMaxWidth()) {
                val min = points.minOf { it.value }
                val max = points.maxOf { it.value }
                Row(Modifier.fillMaxWidth()) {
                    Text(
                        compactNumber(max),
                        style = MaterialTheme.typography.labelSmall,
                        color = colors.textTertiary,
                    )
                }
                Canvas(
                    Modifier
                        .fillMaxWidth()
                        .height(120.dp)
                        .padding(vertical = 4.dp),
                ) {
                    val w = size.width
                    val h = size.height
                    // Bars measure from zero — a bar's whole language is "how
                    // much", and scaling from the series minimum turns the
                    // smallest value into nothing. Lines show shape, so they
                    // stretch over the data's own range.
                    val floor = if (chart.kind == "bar") minOf(min, 0.0) else min
                    val span = (max - floor).takeIf { it != 0.0 } ?: 1.0
                    // A flat series still draws mid-height instead of on the floor.
                    fun yFor(v: Double): Float = (h * (1f - ((v - floor) / span).toFloat() * 0.92f - 0.04f))

                    // Whisper gridlines: top and bottom of the data range.
                    val grid = PathEffect.dashPathEffect(floatArrayOf(6f, 8f))
                    for (frac in listOf(0.04f, 0.96f)) {
                        drawLine(
                            colors.textTertiary.copy(alpha = 0.25f),
                            Offset(0f, h * frac),
                            Offset(w, h * frac),
                            strokeWidth = 1f,
                            pathEffect = grid,
                        )
                    }

                    when (chart.kind) {
                        "bar" -> {
                            val slot = w / points.size
                            val bar = slot * 0.62f
                            val zero = yFor(maxOf(min, 0.0).coerceAtMost(max))
                            points.forEachIndexed { i, p ->
                                val top = yFor(p.value)
                                val left = slot * i + (slot - bar) / 2f
                                val (y0, y1) = if (top <= zero) top to zero else zero to top
                                drawRoundRect(
                                    color = SERIES[i % SERIES.size],
                                    topLeft = Offset(left, y0),
                                    size = Size(bar, maxOf(y1 - y0, 3f)),
                                    cornerRadius = androidx.compose.ui.geometry.CornerRadius(bar * 0.25f),
                                )
                            }
                        }

                        "scatter" -> {
                            points.forEachIndexed { i, p ->
                                val x = w * (if (points.size == 1) 0.5f else i / (points.size - 1f))
                                drawCircle(
                                    color = SERIES[0].copy(alpha = 0.9f),
                                    radius = 7f,
                                    center = Offset(x, yFor(p.value)),
                                )
                            }
                        }

                        else -> { // line and area share the path
                            val path = Path()
                            points.forEachIndexed { i, p ->
                                val x = w * (i / (points.size - 1f))
                                val y = yFor(p.value)
                                if (i == 0) path.moveTo(x, y) else path.lineTo(x, y)
                            }
                            if (chart.kind == "area") {
                                val fill = Path().apply {
                                    addPath(path)
                                    lineTo(w, h)
                                    lineTo(0f, h)
                                    close()
                                }
                                drawPath(
                                    fill,
                                    Brush.verticalGradient(
                                        listOf(SERIES[0].copy(alpha = 0.35f), SERIES[0].copy(alpha = 0.02f)),
                                    ),
                                )
                            }
                            drawPath(path, SERIES[0], style = Stroke(width = 5f))
                            // The "you are here" dot on the last point.
                            val lastY = yFor(points.last().value)
                            drawCircle(SERIES[0], radius = 9f, center = Offset(w, lastY))
                            drawCircle(colors.surface, radius = 4f, center = Offset(w, lastY))
                        }
                    }
                }
                Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        points.first().label.ifBlank { compactNumber(min) },
                        style = MaterialTheme.typography.labelSmall,
                        color = colors.textTertiary,
                        maxLines = 1,
                    )
                    Spacer(Modifier.weight(1f))
                    Text(
                        points.last().label.ifBlank { compactNumber(points.last().value) },
                        style = MaterialTheme.typography.labelSmall,
                        color = colors.textTertiary,
                        maxLines = 1,
                    )
                }
            }
        }
    }
}

@Composable
private fun PieChart(chart: EmbedChart, donut: Boolean, modifier: Modifier = Modifier) {
    val colors = neuColors
    val points = chart.points.filter { it.value.isFinite() && it.value > 0 }.take(8)
    if (points.size < 2) return
    val total = points.sumOf { it.value }.takeIf { it > 0 } ?: return

    Row(modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        Canvas(Modifier.size(110.dp).padding(4.dp)) {
            var start = -90f
            val stroke = size.minDimension * 0.16f
            points.forEachIndexed { i, p ->
                val sweep = (p.value / total * 360.0).toFloat()
                drawArc(
                    color = SERIES[i % SERIES.size],
                    startAngle = start,
                    // A hair of separation, so slices read as slices.
                    sweepAngle = maxOf(sweep - 2f, 1f),
                    useCenter = !donut,
                    style = if (donut) Stroke(width = stroke) else androidx.compose.ui.graphics.drawscope.Fill,
                )
                start += sweep
            }
        }
        Spacer(Modifier.width(14.dp))
        Column(Modifier.weight(1f)) {
            points.forEachIndexed { i, p ->
                Row(verticalAlignment = Alignment.CenterVertically) {
                    androidx.compose.foundation.layout.Box(
                        Modifier.size(8.dp).background(SERIES[i % SERIES.size], CircleShape),
                    )
                    Spacer(Modifier.width(7.dp))
                    Text(
                        p.label.ifBlank { "#${i + 1}" },
                        style = MaterialTheme.typography.labelSmall,
                        color = colors.textSecondary,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.weight(1f, fill = false),
                    )
                    Spacer(Modifier.width(6.dp))
                    Text(
                        "${(p.value / total * 100).toInt()}%",
                        style = MaterialTheme.typography.labelSmall,
                        color = colors.textTertiary,
                    )
                }
                Spacer(Modifier.height(3.dp))
            }
        }
    }
}

/** 118000 -> 118k; keeps axis labels the size of a word. */
private fun compactNumber(v: Double): String {
    val a = abs(v)
    return when {
        a >= 1_000_000 -> trimZero(v / 1_000_000) + "m"
        a >= 1_000 -> trimZero(v / 1_000) + "k"
        v % 1.0 == 0.0 -> v.toInt().toString()
        else -> "%.1f".format(v)
    }
}

private fun trimZero(x: Double): String =
    if (x % 1.0 == 0.0) x.toInt().toString() else "%.1f".format(x)

File: ios/Yappy/Features/Chat/ChartView.swift
import SwiftUI

/// Inline charts for embeds — the whole family, drawn by hand.
///
/// No chart library: a message-sized chart needs exactly one series, no pan,
/// no zoom, no tooltips, and a look that matches the app rather than a
/// spreadsheet. Axis chrome is kept to a whisper (min/max, first/last label)
/// because in a chat bubble the *shape* is the message; anyone who wants the
/// numbers has them in the text beside it.
///
/// Kept in step with android/.../ui/chat/ChartView.kt, which draws the same set.

private let seriesColors: [Color] = [
    Color(hex: 0x8B7CFF), // violet
    Color(hex: 0x00CEC9), // teal
    Color(hex: 0xFF9F43), // orange
    Color(hex: 0xFF6B81), // pink
    Color(hex: 0xFCCE09), // yellow
    Color(hex: 0x6BCB77), // green
    Color(hex: 0x4FC3F7), // sky
    Color(hex: 0xB39DDB), // lavender
]

struct ChartView: View {
    @Environment(\.neu) private var colors

    let chart: EmbedChart

    var body: some View {
        let points = chart.points.filter { $0.value.isFinite }
        if points.count >= 2 {
            if chart.kind == "pie" || chart.kind == "donut" {
                PieChartView(chart: chart, donut: chart.kind == "donut")
            } else {
                linear(points)
            }
        }
    }

    private func linear(_ points: [EmbedChartPoint]) -> some View {
        let minValue = points.map(\.value).min() ?? 0
        let maxValue = points.map(\.value).max() ?? 0

        return VStack(alignment: .leading, spacing: 0) {
            Text(compactNumber(maxValue))
                .font(YappyFont.labelSmall)
                .foregroundStyle(colors.textTertiary)

            Canvas { context, size in
                draw(points, minValue: minValue, maxValue: maxValue, in: &context, size: size)
            }
            .frame(height: 120)
            .padding(.vertical, 4)

            HStack(spacing: 0) {
                Text(firstLabel(points, minValue: minValue))
                    .font(YappyFont.labelSmall)
                    .foregroundStyle(colors.textTertiary)
                    .lineLimit(1)
                Spacer(minLength: 8)
                Text(lastLabel(points))
                    .font(YappyFont.labelSmall)
                    .foregroundStyle(colors.textTertiary)
                    .lineLimit(1)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func firstLabel(_ points: [EmbedChartPoint], minValue: Double) -> String {
        let label = points[0].label
        return label.isEmpty ? compactNumber(minValue) : label
    }

    private func lastLabel(_ points: [EmbedChartPoint]) -> String {
        let last = points[points.count - 1]
        return last.label.isEmpty ? compactNumber(last.value) : last.label
    }

    private func draw(
        _ points: [EmbedChartPoint],
        minValue: Double,
        maxValue: Double,
        in context: inout GraphicsContext,
        size: CGSize
    ) {
        let w = size.width
        let h = size.height

        // Bars measure from zero — a bar's whole language is "how much", and
        // scaling from the series minimum turns the smallest value into
        // nothing. Lines show shape, so they stretch over the data's own range.
        let floorValue = chart.kind == "bar" ? Swift.min(minValue, 0) : minValue
        let rawSpan = maxValue - floorValue
        let span = rawSpan == 0 ? 1 : rawSpan
        // A flat series still draws mid-height instead of on the floor.
        func yFor(_ v: Double) -> CGFloat {
            h * (1 - CGFloat((v - floorValue) / span) * 0.92 - 0.04)
        }

        // Whisper gridlines: top and bottom of the data range.
        for frac in [0.04, 0.96] as [CGFloat] {
            var line = Path()
            line.move(to: CGPoint(x: 0, y: h * frac))
            line.addLine(to: CGPoint(x: w, y: h * frac))
            context.stroke(
                line,
                with: .color(colors.textTertiary.opacity(0.25)),
                style: StrokeStyle(lineWidth: 1, dash: [3, 4])
            )
        }

        switch chart.kind {
        case "bar":
            let slot = w / CGFloat(points.count)
            let bar = slot * 0.62
            let zero = yFor(Swift.min(Swift.max(minValue, 0), maxValue))
            for (i, p) in points.enumerated() {
                let top = yFor(p.value)
                let left = slot * CGFloat(i) + (slot - bar) / 2
                let (y0, y1) = top <= zero ? (top, zero) : (zero, top)
                let rect = CGRect(x: left, y: y0, width: bar, height: Swift.max(y1 - y0, 2))
                context.fill(
                    Path(roundedRect: rect, cornerRadius: bar * 0.25, style: .continuous),
                    with: .color(seriesColors[i % seriesColors.count])
                )
            }

        case "scatter":
            for (i, p) in points.enumerated() {
                let x = points.count == 1
                    ? w * 0.5
                    : w * CGFloat(i) / CGFloat(points.count - 1)
                let y = yFor(p.value)
                context.fill(
                    Path(ellipseIn: CGRect(x: x - 3, y: y - 3, width: 6, height: 6)),
                    with: .color(seriesColors[0].opacity(0.9))
                )
            }

        default: // line and area share the path
            var path = Path()
            for (i, p) in points.enumerated() {
                let point = CGPoint(
                    x: w * CGFloat(i) / CGFloat(points.count - 1),
                    y: yFor(p.value)
                )
                if i == 0 { path.move(to: point) } else { path.addLine(to: point) }
            }

            if chart.kind == "area" {
                var fill = path
                fill.addLine(to: CGPoint(x: w, y: h))
                fill.addLine(to: CGPoint(x: 0, y: h))
                fill.closeSubpath()
                context.fill(
                    fill,
                    with: .linearGradient(
                        Gradient(colors: [seriesColors[0].opacity(0.35), seriesColors[0].opacity(0.02)]),
                        startPoint: .zero,
                        endPoint: CGPoint(x: 0, y: h)
                    )
                )
            }

            context.stroke(
                path,
                with: .color(seriesColors[0]),
                style: StrokeStyle(lineWidth: 2, lineCap: .round, lineJoin: .round)
            )

            // The "you are here" dot on the last point.
            let lastY = yFor(points[points.count - 1].value)
            context.fill(
                Path(ellipseIn: CGRect(x: w - 3.5, y: lastY - 3.5, width: 7, height: 7)),
                with: .color(seriesColors[0])
            )
            context.fill(
                Path(ellipseIn: CGRect(x: w - 1.5, y: lastY - 1.5, width: 3, height: 3)),
                with: .color(colors.surface)
            )
        }
    }
}

/// Pie and donut: arcs on the left, a legend on the right, because at this
/// size labels *on* the slices would be unreadable.
private struct PieChartView: View {
    @Environment(\.neu) private var colors

    let chart: EmbedChart
    let donut: Bool

    var body: some View {
        let points = Array(chart.points.filter { $0.value.isFinite && $0.value > 0 }.prefix(8))
        let total = points.reduce(0.0) { $0 + $1.value }

        if points.count >= 2, total > 0 {
            HStack(spacing: 14) {
                Canvas { context, size in
                    let radius = min(size.width, size.height) / 2
                    let center = CGPoint(x: size.width / 2, y: size.height / 2)
                    let strokeWidth = min(size.width, size.height) * 0.16
                    var start: Double = -90

                    for (i, p) in points.enumerated() {
                        let sweep = p.value / total * 360
                        // A hair of separation, so slices read as slices.
                        let drawn = Swift.max(sweep - 2, 1)
                        let color = seriesColors[i % seriesColors.count]

                        // In the flipped coordinate space `clockwise: false`
                        // sweeps clockwise on screen, matching a positive
                        // Android sweepAngle.
                        if donut {
                            var arc = Path()
                            arc.addArc(
                                center: center,
                                radius: radius - strokeWidth / 2,
                                startAngle: .degrees(start),
                                endAngle: .degrees(start + drawn),
                                clockwise: false
                            )
                            context.stroke(arc, with: .color(color), style: StrokeStyle(lineWidth: strokeWidth))
                        } else {
                            var wedge = Path()
                            wedge.move(to: center)
                            wedge.addArc(
                                center: center,
                                radius: radius,
                                startAngle: .degrees(start),
                                endAngle: .degrees(start + drawn),
                                clockwise: false
                            )
                            wedge.closeSubpath()
                            context.fill(wedge, with: .color(color))
                        }
                        start += sweep
                    }
                }
                .frame(width: 102, height: 102)
                .padding(4)

                VStack(alignment: .leading, spacing: 3) {
                    ForEach(Array(points.enumerated()), id: \.offset) { i, p in
                        HStack(spacing: 0) {
                            Circle()
                                .fill(seriesColors[i % seriesColors.count])
                                .frame(width: 8, height: 8)
                            Text(p.label.isEmpty ? "#\(i + 1)" : p.label)
                                .font(YappyFont.labelSmall)
                                .foregroundStyle(colors.textSecondary)
                                .lineLimit(1)
                                .padding(.leading, 7)
                            Spacer(minLength: 6)
                            Text("\(Int(p.value / total * 100))%")
                                .font(YappyFont.labelSmall)
                                .foregroundStyle(colors.textTertiary)
                        }
                    }
                }
                .frame(maxWidth: .infinity)
            }
            .frame(maxWidth: .infinity)
        }
    }
}

/// 118000 -> 118k; keeps axis labels the size of a word.
private func compactNumber(_ v: Double) -> String {
    let a = abs(v)
    if a >= 1_000_000 { return trimZero(v / 1_000_000) + "m" }
    if a >= 1_000 { return trimZero(v / 1_000) + "k" }
    if v.truncatingRemainder(dividingBy: 1) == 0 { return String(Int(v)) }
    return String(format: "%.1f", v)
}

private func trimZero(_ x: Double) -> String {
    x.truncatingRemainder(dividingBy: 1) == 0 ? String(Int(x)) : String(format: "%.1f", x)
}
